### PR TITLE
Expose HttpSys RequestInfo extensibility 

### PR DESCRIPTION
**Pull request by Tratcher**
#13482 There are asks to expose data from the Http.Sys RequestInfo extensibility point. However, many of the extensions are still under development and don't meet our documentation requirements. We can mitigate this by exposing the raw bytes and letting the caller interpret them as needed.

### CODE REVIEW

File: src/Servers/HttpSys/test/FunctionalTests/HttpsTests.cs [hunk 1]
```diff
@@ -160,6 +163,46 @@ public async Task Https_SetsITlsHandshakeFeature()
...
+                    Assert.True(requestInfoFeature.RequestInfo.Count > 0);
+                    var tlsInfo = requestInfoFeature.RequestInfo.Where(info => info.Key == (int)HttpApiTypes.HTTP_REQUEST_INFO_TYPE.HttpRequestInfoTypeSslProtocol).Single().Value;
+                    HttpApiTypes.HTTP_SSL_PROTOCOL_INFO tlsCopy;
+                    unsafe
+                    {
+                        using var handle = tlsInfo.Pin();
+                        tlsCopy = Marshal.PtrToStructure<HttpApiTypes.HTTP_SSL_PROTOCOL_INFO>((IntPtr)handle.Pointer);
+                    }
+
+                    // Assert.Equal(tlsFeature.Protocol, tlsCopy.Protocol); // These don't directly match because there's some value mapping that needs to happen.
```
Comment (jkotalik) [inline src/Servers/HttpSys/test/FunctionalTests/HttpsTests.cs]: Clarify on this?

File: src/Shared/HttpSys/RequestProcessing/NativeRequestContext.cs [hunk 2]
```diff
@@ -478,5 +479,42 @@ private uint GetChunksHelper(ref int dataChunkIndex, ref uint dataChunkOffset, b
...
+                {
+                    var request = (HttpApiTypes.HTTP_REQUEST_V2*)(pMemoryBlob + _bufferAlignment);
+                    return GetRequestInfo(_originalBufferAddress, request);
+                }
+            }
+        }
+
+        private IReadOnlyCollection<KeyValuePair<int, ReadOnlyMemory<byte>>> GetRequestInfo(IntPtr baseAddress, HttpApiTypes.HTTP_REQUEST_V2* nativeRequest)
+        {
+            if (nativeRequest->RequestInfoCount == 0)
```
Comment (jkotalik) [inline src/Shared/HttpSys/RequestProcessing/NativeRequestContext.cs]: Nit: probably faster to save nativeRequest->RequestInfoCount in a local

File: src/Servers/HttpSys/test/FunctionalTests/HttpsTests.cs [hunk 3]
```diff
@@ -160,6 +163,46 @@ public async Task Https_SetsITlsHandshakeFeature()
...
+            using (Utilities.CreateDynamicHttpsServer(out var address, async httpContext =>
+            {
+                try
+                {
+                    var tlsFeature = httpContext.Features.Get<ITlsHandshakeFeature>();
+                    var requestInfoFeature = httpContext.Features.Get<IHttpSysRequestInfoFeature>();
+                    Assert.NotNull(tlsFeature);
+                    Assert.NotNull(requestInfoFeature);
+                    Assert.True(requestInfoFeature.RequestInfo.Count > 0);
+                    var tlsInfo = requestInfoFeature.RequestInfo.Where(info => info.Key == (int)HttpApiTypes.HTTP_REQUEST_INFO_TYPE.HttpRequestInfoTypeSslProtocol).Single().Value;
```
Comment (halter73) [inline src/Servers/HttpSys/test/FunctionalTests/HttpsTests.cs]: Nit: you can just pass the Where predicate to Single() and skip the Where.

File: src/Servers/HttpSys/test/FunctionalTests/HttpsTests.cs (hunk 1 above)
Comment (Tratcher) [reply src/Servers/HttpSys/test/FunctionalTests/HttpsTests.cs]: The managed and native enum values don't directly align.
https://github.com/aspnet/AspNetCore/blob/921816263779c158e3d1d76591d63d928cc9c669/src/Servers/HttpSys/src/RequestProcessing/Request.cs#L260-L285

File: src/Shared/HttpSys/RequestProcessing/NativeRequestContext.cs (hunk 2 above)
Comment (Tratcher) [reply src/Shared/HttpSys/RequestProcessing/NativeRequestContext.cs]: Cleaner, but not faster.

Comment (jkotalik) [reply src/Shared/HttpSys/RequestProcessing/NativeRequestContext.cs]: yeah cause pointers, not properties.

File: src/Shared/HttpSys/RequestProcessing/NativeRequestContext.cs [hunk 4]
```diff
@@ -478,5 +480,43 @@ private uint GetChunksHelper(ref int dataChunkIndex, ref uint dataChunkOffset, b
             }
             return dataRead;
         }
+
+        internal IReadOnlyDictionary<int, ReadOnlyMemory<byte>> GetRequestInfo()
+        {
+            if (_permanentlyPinned)
```
Comment (jkotalik) [inline src/Shared/HttpSys/RequestProcessing/NativeRequestContext.cs]: How frequently do both codepaths get hit here? What is the default setting for pinning?

**Review by jkotalik — APPROVED** [review]

File: src/Shared/HttpSys/RequestProcessing/NativeRequestContext.cs (hunk 4 above)
Comment (Tratcher) [reply src/Shared/HttpSys/RequestProcessing/NativeRequestContext.cs]: _permanentlyPinned is for IIS, HttpSys always uses the other one.

Comment (jkotalik) [reply src/Shared/HttpSys/RequestProcessing/NativeRequestContext.cs]: Oh yeah... forgot I added that XD

**Comment by Tratcher** [comment]
@anurse what was that about tactics approval? This should have been on @shirhatti's pre-approval list.

**Comment by anurse** [comment]
If that's the case, we should be fine.

**Pull request merged**